pkg/status/status.go

```diff
@@ -285,16 +285,7 @@ func (s *status) monitorClusterStatus() {
 	for {
 		select {
 		case <-s.stopCh:
-			// If the stopCh is closed, the operator will exit and CO should
-			// be set to degraded.
-			conditionListBuilder := clusterStatusListBuilder()
-			conditionListBuilder(configv1.OperatorProgressing, configv1.ConditionFalse, "")
-			conditionListBuilder(configv1.OperatorAvailable, configv1.ConditionFalse, "")
-			statusConditions := conditionListBuilder(configv1.OperatorDegraded, configv1.ConditionTrue, "Operator exited")
```

Comment:
It will be useful if we updated the `Message` to reflect what happened. Example: `Operator exited with status...`. I am also worried that we are not looking at the exit values here and an actual issue will be ignored.

Reply:
@aravindhp
>It will be useful if we updated the Message to reflect what happened. Example: Operator exited with status...

That's possible, but we will need to update how we listen for `SIGTERM` and `SIGINT`. 

We currently [start the manager](https://github.com/operator-framework/operator-marketplace/blob/master/cmd/manager/main.go#L104) using OLM's [signals package](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/lib/signals/signals.go) to and will need to reproduce the function if we want to modify it. Thoughts?

Reply:
Also - where should we track the message - are you suggesting that we report `degraded` as false but update its message?

Reply:
>We currently start the manager using OLM's signals package to and will need to reproduce the function if we want to modify it. Thoughts?

If changes are needed we can copy the function over and modify it. You can also talk to the OLM folks about it to check if they are facing a similar issue.

>Also - where should we track the message - are you suggesting that we report degraded as false but update its message?

If I look at the cluster version page on the UI it should be reflected there and from the CLI. I leave it to you as to where it needs to be updated.

Reply:
> If changes are needed we can copy the function over and modify it. You can also talk to the OLM folks about it to check if they are facing a similar issue.

After investigating how to report the exit code to the cluster operator I don't believe that we should report the exit code in the cluster operator status. Instead, we can simply log the exit code in the fatal function and reflect that the operator has exited in the cluster operator status like so:

Updated fatal code:
```
// exit updates the clusteroperator status and exits with the correct error code.
func exit(err error, monitorDoneCh <-chan struct{}, monitorStopCh chan struct{}) {
	// Stop reporting clusterOperator status
	close(monitorStopCh)

	// Wait for clusterOperator status reporting to stop
	<-monitorDoneCh

	// If an error exists then exit with status set to 1
	if err != nil {
		log.Fatalf("The operator encountered an error, exit code 1: %v", err)
	}

	// No error, graceful termination
	log.Info("The operator exited gracefully, exit code 0")
	os.Exit(0)
}
```

The `CO` status should be updated so degraded is not set to true and to update the available message to let the admin know that it is no longer available:
```
Available = false, "The operator has exited and is no longer reporting status."
Progressing = false, ""
Degraded = false, ""
```
This approach will let CVO know that marketplace is no longer available and give admins a full idea of what happen in the logs while introducing minimal changes.

Reply:
>Example updated fatal code

What is the point in just logging? You are still not checking to see what the signal was. Moreover you are giving a false negative (in the error message) in the case of SIGTERM and false positive (in the CO) in the case of an actual bad exit.

>This approach will let CVO know that marketplace is no longer available

How will updating the message inform the CVO that the operator is not available?

Reply:
> > After investigating how to report the exit code to the cluster operator I don't believe that we should report the exit code in the cluster operator status.
> 
> Please explain why.
> 
> > Example updated fatal code
> 
> What is the point in just logging? You are still not checking to see what the signal was. Moreover you are giving a false negative (in the error message) in the case of SIGTERM and false positive (in the CO) in the case of an actual bad exit.
> 
> > This approach will let CVO know that marketplace is no longer available
> 
> How will updating the message inform the CVO that the operator is not available?

I'm not sure I totally agree here. It seems like the purpose of sending ClusterOperator status is to describe the current state of the operator. Can we define these terms to maybe better help come up with what is "accurate"? My imagination of these terms:

`Available`: The operator is responding to requests in the controller
`Progressing`: The operator is coming up and is not yet ready to respond to requests
`Degraded`: The operator, due to some present issue, is unable to successfully respond to requests.

I don't think any of these fields describe the state of the operator that is "not running". It seems like the correct solution, in that case, is on exit just mark the cluster status field `Available` to false. I don't think that an exit code of `0` vs an exit code of `1` would result in a different combination here. And, if that is the case, then simply logging the exit code in the marketplace operator should be sufficient here. What value do we get by adding something to the cluster operator message that can't be gained by just looking at the marketplace operator's logs? What is the goal of the cluster status other than a high level aggregation of state?

I'm not sure I see what "false positive" we would be sending using Alex's implementation description above. Is there something I'm mising here?

Reply:
>Please explain why.

Kevin seems to have beaten me to it.

>Moreover you are giving a false negative (in the error message) in the case of SIGTERM

`SIGTERM` wouldn't enter the `if statement` because the fatal function would be called with a `nil` error, so we would update the cluster operator as unavailable and exit with code 0. I tested this by creating a deployment with the changes and killing the pod.

I do not see the false positive case either.

Reply:
>I do not see the false positive case either.

```
Available = false, "The operator has exited and is no longer reporting status."
Progressing = false, ""
Degraded = false, ""
```
In the case of a real failure shouldn't `degraded` be set to `true`?
Also, is setting all conditions to `false` allowed?

>What value do we get by adding something to the cluster operator message that can't be gained by just looking at the marketplace operator's logs?

We don't want the cluster admin to have to go look at the operator logs to figure out what the problem is in the case of failure. That is exactly what the ClusterOperator CR is for.

Reply:
> In the case of a real failure shouldn't degraded be set to true?

I don't believe this is the case given these descriptions found [here](https://godoc.org/github.com/openshift/api/config/v1#ClusterStatusConditionType)

```
Available indicates that the binary maintained by the operator 
(eg: openshift-apiserver for the openshift-apiserver-operator), is 
functional and available in the cluster.
```
VS.
```
Degraded indicates that the operand is not functioning completely. 
An example of a degraded state would be if there should be 5 copies 
of the operand running but only 4 are running. It may still be available,
but it is degraded
```

> Also, is setting all conditions to false allowed?

Based on the descriptions available in the link  above and the [official ClusterOperator docs](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md) there is no reason why they can't given that they each describe separate states. I can reach out to confirm.

> We don't want the cluster admin to have to go look at the operator logs to figure out what the problem is in the case of failure. That is exactly what the ClusterOperator CR is for.

I agree that this is the intent of the CR. We can add this to the CR message but it will require that we convert the [monitorStopCh](https://github.com/operator-framework/operator-marketplace/blob/master/pkg/status/status.go#L287) to take a variable. This isn't a bad thing - but I had wanted to remove the monitorStopCh and replace it with a single channel that our operator could use to close all child processes. It's a nice to have but not breaking.

Reply:
I also don't believe that asking a cluster administrator to inspect the logs is an unreasonable ask.

Reply:
> I also don't believe that asking a cluster administrator to inspect the logs of an unavailable operator is an unreasonable ask.

Not if it can be helped. It is just a bad user experience.

Reply:
I just think that Available=False describes the state of the cluster. I don't believe that some combination of these states accurately describes the difference between "stopped because we asked nicely" and "stopped rudely". That distinction seems arbitrary from the perspective of cluster status, so why spend significant engineering time trying to solve it? That's my take here.

Reply:
I am okay with having all three conditions set to false with an appropriate message as long as that's allowed.

Reply:
There seems to be no reason why we cannot after reviewing the following document [1] [2] and checking on Slack to confirm.

[1] https://godoc.org/github.com/openshift/api/config/v1#ClusterStatusConditionType
[2] https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md
[3] https://coreos.slack.com/archives/CEKNRGF25/p1557174280220900